### DISCUSSION TITLE
Discovering nested types inside record primary constructor

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6286] [Language Ideas]
Currently, within the primary constructor of a record, its nested types are not usable without qualifying them with the record type. Consider the following example:
```csharp
public record Record(Option SomeOption)
{
    public enum Option { Right, Wrong }
}
```

This will report a CS0246 error (`The type or namespace name 'Option' could not be found (are you missing a using directive or an assembly reference?)`). You have to use the record name to qualify the type, meaning the following becomes legal:
```csharp
public record Record(Record.Option SomeOption)
{
    public enum Option { Right, Wrong }
}
```

I think this can be improved, removing the need to qualify the type via its container record type, making the above example legal.

It is noteworthy that the nested types are still visible, meaning you can have the following:
```csharp
public record Record
{
    internal record NestedRecord(NestedRecord.Option SomeOption)
    {
        internal enum Option { Right, Wrong }
    }
}
```

**Comment by FaustVX** [comment]
It look like the inheritance
```c#
public class C : C.I // C : I  <- will throw the same error here
{
  public interface I { }
}
```

**Comment by TahirAhmadov** [comment]
This rule existed since C# 1, and I'm pretty sure there was a reason why it works like this, to do with name resolution; this is not specific to records. Most likely, once you allow that resolution rule, it opens a can of worms, as to what the exact rules are, etc. Also, it's most likely a breaking change:
```cs
class Foo { }
class Bar : Foo // this currently refers to outside Foo, but with this change, it silently starts referring to the inside Foo
{
  class Foo { }
}
```

**Reply by Rekkonnect** [reply]
It bothers me because the primary constrctor is still a constructor, and I would expect it to behave similarly to if I were to declare it inside the body of the record.

Plus, avoid naming collisions anyway. Having Foo and Bar.Foo very sounds like a *you* problem if it breaks.

**Reply by TahirAhmadov** [reply]
I see what you mean. It's one of those choices - you have two possible approaches, each one consistent with a different rule, depending on whether you look at it as a constructor, vs as a class definition. Maybe it's possible to make a limited change for records only, but I'm still unsure whether the possible breaking change would be justified. 
And I agree - naming collisions like that are most likely a problem in the code base; however, C# has to try hard not to break even bad code.

**Reply by Rekkonnect** [reply]
They've taken the brave step to breaking delegates in some areas, fixing scoping in that very niche scenario doesn't sound like something people would whine about. Worst case scenario, you get two unexpected errors in your entire solution per 158 users*, I wouldn't mind. 

*Figures were ass-pulled, and no accountability for their exactness is held
